test-project/composer.json

```diff
@@ -7,7 +7,12 @@
             "type": "composer",
             "url": "http://localhost:8080",
             "tuf": {
-                "root": "root.json"
+                "root": {
+                    "hashes": {
+                        "sha256": "6cf95b77cedc832c980b81560704bd2fb9ee32ec4c1a73395a029b76715705cc"
```

Comment:
If I remember correctly, the TUF spec locates the root key offline (e.g., literally in a physical safe or something) and suggests rotating it roughly annually. Wouldn't this sha hash change when that happens? And in that case, would this "just stop working" one day for our users until they update the hash from the webpage or whatever they originally got it from? cc @davidstrauss @tedbow

Reply:
I'd have to review the code again to understand how it's implemented, but what it *should* do is only use this pinning to bootstrap new clients. The code should probably ship with the version for that, too.

When we rotate root, it should have the following effects with this design:

- No special effect on bootstrapped clients, which have already pulled and stored local root data and no longer use the bootstrapping information.
- If someone installs using a download from before the rotation, their client will need to bootstrap to the old root and update to the new one by standard means. This would also be the case if someone installed from a stale tarball that included the full but stale root metadata. We're required to keep the chain of rotated root data published, at least as far back as we expect clients to need to bootstrap.

Here's how my thought process went around working with this root bootstrapping problem:

1. We could use a design where we ship the initial root data in a way that both populates the client's repo metadata store and retains its writability, but that's kind of awkward. We usually like to split things into (a) shipped and nearly immutable vs. (b) mutable but initially empty (e.g. /files).
2. We could ship root.json and copy it into the client repo metadata. It's kind of unwieldly to embed in other JSON, though. We'd probably want to ship it separately and reference it somehow in the repo config.
3. Since old repo metadata must remain published, we ought to be able to bootstrap with just a hash (and maybe size) for the data. If it can't fetch the current (or fresher) root data, then it probably can't pull anything else from the repo. Having a full root.json doesn't seem to help the client survive any attacks above and beyond providing it the hash + size.

Still, I'll run this past people at the Secure Systems Lab.

Reply:
From Marina Moore on TUF Slack:

> For the purpose of bootstrapping trust, the secure hash should be functionally equivalent to the root metadata, so that seems like a reasonable space-saving mechanism. It also gives more paranoid users an opportunity to look for external verification that they have the correct root metadata.

Reply:
I am unclear of the motivation for this PR. @phenaproxima in the first comment describes the problem as "we can't easily resolve the path of the starting root metadata,"

but then @davidstrauss in TUF slack mentions "include it in Composer's repository configuration in a compact way" and Marina Moore mentions "reasonable space-saving mechanism". 

Is there not an actual problem of resolving? Are the initial root.json files so big we are worried about the space?

Reply:
It's just awkward to have the Composer config either (1) reference a file elsewhere on disk or (2) embed the root data into an otherwise sensible JSON file.

Another option could be some sort of naming convention for discovering the root data for a given repo. For example, we could use the FQDN or hash of the base URL. This would allow populating the root on disk without awkwardly referencing it from the JSON.

Reply:
I'm not sure that would solve the problem. The issue is, if we expect to find the root JSON in the file system, we need to know its path. And as you know, there are two kinds of paths: relative and absolute. Both present problems:

If we used a relative path, then what is it relative to? The current working directory? The plugin's `src` directory? The composer.json file itself? Some package's path in the vendor directory? (If it's relative to a package, what's the path of the package? We cannot know this until the package info is loaded, but TUF needs to be fully bootstrapped and updated before we can even begin to figure that out.)

If it's an absolute URL, then how can we guarantee that path be valid in all environments the composer.json file might live in?

Downloading the root JSON is not my favorite idea, but it _does_ get around these questions, because the root file lives at a stable, known location (i.e., a URL on the server). But I'm absolutely open to other ideas.

Reply:
One option that @davidstrauss suggested (somewhere...can't remember if it's in this issue or some other one) is that, if say, we want people to add TUF protection to the `packages.drupal.org` repository, we document that we expect the project to have a directory structure like this:

```
.pki/
    tuf/
        packages.drupal.org.json
composer.json
```